File: leetcode/1046. Last Stone Weight/main.go
package main

import (
	"sort"
)

func IntAbs(x, y int) int {
	ans := x - y
	if ans < 0 {
		return -ans
	}

	return ans
}

func lastStoneWeight(stones []int) int {
	for len(stones) > 1 {
		sort.Ints(stones)
		x, y := stones[len(stones)-1], stones[len(stones)-2]
		if x == y {
			stones = stones[:len(stones)-2]
			continue
		} else {
			stones = stones[:len(stones)-2]
			stones = append(stones, IntAbs(x, y))
			continue
		}
	}

	if len(stones) == 0 {
		return 0
	}

	return stones[0]
}

File: leetcode/1046. Last Stone Weight/main.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn last_stone_weight(stones: Vec<i32>) -> i32 {
        let mut pq = BinaryHeap::from(stones);

        loop {
            match pq.len() {
                0 => return 0,
                1 => return pq.pop().unwrap(),
                _ => {}
            }

            let x = pq.pop().unwrap();
            let y = pq.pop().unwrap();

            if x != y {
                pq.push(i32::abs(x - y));
            }
        }
    }
}